Refactor OAuth to pass an external implementation

woehrl01 opened the pull request:
PR to be able to replace the OAuth dealing with external logics, e.g. Home Assistant (#115)

Changes:

* Added abstract implementation which wraps the local OAuth handling
* Added a new facade class `PyViCare` to interact with your devices. With this class, you can iterate over all your existing devices. See the changed README for usage description
* Added `ViCareHomeAssistantOAuthManager` as an example, which of course should be moved to the Home Assistant part.
* Added name of the device and online status output.
* Added an integration test which automatically runs against your local device, you need to provide credentials for this. The tests can be run with `./run_test.sh` (you likely need to `chmod +x` first)
* Added a `./format.sh` script to have a consistent formatting over the code. Will create a PR after this one is merged.
* Added an `PyViCareInvalidCredentialsError` which gets thrown if there are invalid credentials provided and a token can't be fetched
* Some other minor refactorings and bugfixes, I came along throughout the major refactoring and can be grouped together.

woehrl01 committed:
Extract OAuth handling into separate class

woehrl01 committed:
Added factory to create the service

woehrl01 committed:
refactored to builder to get rid of optional args

woehrl01 committed:
typo

woehrl01 committed:
remove optional arg

woehrl01 committed:
run prettier

woehrl01 committed:
Only use cached service on positive cacheduration

woehrl01 committed:
pass correct curcuit to builder in tests

crazyfx1 reviewed (commented):
The refactoring looks great!

I don't like the new interface though. I don't want the user to deal with the service instance.

In an ideal world I would imagine the usage somehow like this:
There is an PyViCare object with an init method to initialize the connection. Then everything is already filled.
```Python
pyvicare = PyViCare()
pyvicare.init(username, password, clientid)
# or 
pyvicare.init(oauthmanager) 
# pyvicare object is now populated
pyvicare.installations # contains installation
pyvicare.installations[0].gateways # contains gateways
pyvicare.installations[0].gateways[0].devices # contains devices
pyvicare.devices # contains all devices directly aswell
pyvicare.devices[0] # is already of type GazBoiler if you have a GazBoiler
```

Unfortunately we can't really tell which device we use, so the last step is not possible. But we could just implement
```Python
pyvicare.getGazBoilerDevice(0) # gets first device which user has to know is a gas boiler
```

For this to work the init method must already execute one data request to the API. But since oischinger wants to check which sensors are available on initialization we need that call anyways ;)

crazyfx1 commented:
Ok you now changed to a builder pattern, which works ofc. I can imagine that a level higher:
```Python
pyvicare = PyViCareBuilder()
                     .withClientId(clientId)
                     .withCredentials(username, password)
                     .build()
```

woehrl01 commented:
Thanks for your input. I guess we are on a good track for this, with the current implementation. I suggest creating a `PyViCare` class which simple wraps the builder and all the other types. 
Furthermore, I had the idea of adding the type to the builder, too. But that hasn't looked right. Your idea filled the gap, awesome.

edit: I just added an initial version of a `PyViCare` class. To make this work nicely, I need to do some heavier refactoring, so we can pass the serial, installation and circuit around more freely.
edit: done.

woehrl01 committed:
Move exceptions and other types to Util file

woehrl01 committed:
implement initial pyvicare facade

woehrl01 committed:
added intermediate class to store access combination

woehrl01 committed:
remove builder and replace with PyViCare facade

woehrl01 committed:
update readme

woehrl01 committed:
no implicit idx

woehrl01 committed:
iterate all devices

woehrl01 commented on PyViCare/PyViCare.py:
```diff
@@ -1,42 +1,67 @@
+            for gateway in installation["gateways"]:
+                serial = gateway["serial"]
+
+                accessor = ViCareDeviceAccessor(id, serial, 0)
```
We still miss iterating through the devices.

woehrl01 commented on PyViCare/PyViCareOAuthManager.py:
```diff
@@ -0,0 +1,215 @@
+            return self.post(url, data)
+
+
+class ViCareHomeAssistantOAuthManager(AbstractViCareOAuthManager):
```
Move this to the HomeAssistant implementation

woehrl01 committed:
itereate through all heating devices

woehrl01 committed:
provide config as iterable argument

woehrl01 commented:
@crazyfx1 Please have another look, I implemented the iteration over all devices. 

I thought about replicating the installations and gateways, but I don't know about a reasonable use case for that. As we are only accessing the devices, I created a list of all devices, like in your pseudocode. Each device has all necessary information stored in the `.config` property, for filtering, I'd suggest using that data.

woehrl01 committed:
format

woehrl01 committed:
Add device model and status

crazyfx1 commented:
I like the new interface very much. I wanted to try it out but I couldn't get it imported correctly. Maybe it's because the class is also called PyViCare just like the library and the module?

woehrl01 committed:
rename to ViCare

woehrl01 commented:
Just renamed it to `ViCare` for now, not sure about if we get licensing issues here?

edit: found the issue. I'm about to fix it

crazyfx1 commented:
Did you find the circular import?

woehrl01 committed:
fix cyclic dependecy and added integration test you can simply run to test aagainst your own device

woehrl01 commented:
So, I fixed the cyclic dependency and added an integration test you can run with your own credentials. That means we are always on the safe spot for the new `PyViCare` class. The name is changed back to `PyViCare`, as it's not an issue.

woehrl01 committed:
rename test

woehrl01 committed:
move stuff around, add comments

woehrl01 committed:
fix import

crazyfx1 commented:
![image](https://user-images.githubusercontent.com/9930385/127027818-06028d4f-de9c-43df-b71c-de762d81dec6.png)

The PyViCareCachedService is now broken :)

woehrl01 commented:
😅 On my way... 

edit: done

woehrl01 committed:
fix CachedService and add generic output to integration test

crazyfx1 commented:
Ok the access works. Only the properties with service.circuit in it are broken :)

woehrl01 committed:
fix access to circuit id and raise exception if bad credentials provided

woehrl01 committed:
add format script

woehrl01 commented:
Thank you very much @crazyfx1 for testing this so in-depth on your side. I fixed the `circuit` reference. I also just stumbled over the issue if you have a misconfiguration of credentials. I added a meaningful exception in that case.

I also added a `format.sh` file, which we can run after everything, which should give us some cleaning. I didn't run it yet as it would bloat the diff pretty hard.

woehrl01 committed:
update readme

woehrl01 committed:
Add output for online status to integration test

woehrl01 committed:
Run through all getter methods for integration test

woehrl01 committed:
add output for skip

woehrl01 committed:
fix typo, preserve signatures for decorated functions, add logging

woehrl01 committed:
refactored token re-/store logic in oauth manager

woehrl01 committed:
format integration test output

woehrl01 committed:
pretty print dictionary

woehrl01 committed:
configure device type for integration

woehrl01 committed:
small tweaks about device type passing

woehrl01 committed:
improve readability

woehrl01 committed:
Implement autodetection of device type based on device name

woehrl01 committed:
switch to generator

woehrl01 committed:
refactor device detection

woehrl01 committed:
Add test for device config

woehrl01 committed:
fix missing return statement

woehrl01 committed:
remove unused code

oischinger commented:
Oh boy, this is quite a huge refactoring.
Meanwhile I've started a config flow for the HA integration here: https://github.com/oischinger/ha_vicare/pull/43

This requires the classic credentials and client_id but we can modify it to use the HA oauth flow steps.

woehrl01 commented:
@oischinger the refactoring looks bigger than it is. Most of the stuff is just moved around to other files. The testing part is a complete addon. The addition of the PyViCare class was needed to be able to replace the oauthmanager in a seemless way. It should make the usage from Home Assistant much easier, as we can now automatically detect all devices and also the device type based on the name of the device.

woehrl01 committed:
Add readme
